keps/sig-storage/20190120-execution-hook-design.md

```diff
@@ -0,0 +1,546 @@
+actions when taking a snapshot. There is an existing lifecycle hook in the `Container`
+struct. The lifecycle hook is called immediately after a container is created or
+immediately before a container is terminated. The proposed execution hook is not
+tied to the start or termination time of the container. Instead it is called before
+or after taking a snapshot while the container is still running.
+
+### Goals
+
+`ExecutionHook` is introduced to define actions that can be taken on a container or a group of containers. The ExecutionHook controller is responsible for triggering the commands in containers and updating the status on whether the execution is succeeded or not. The controller will also garbage collect the hook objects based on some predefined policy.
+
+### Non-Goals
+
+This proposal does not provide exact command included in the `ExecutionHook`
+because every application has a different requirement.
+
+ApplicationSnapshot and GroupSnapshot will be mentioned in this proposal whenever relevant, but detailed design will not be included in this spec.
+
+## Proposal
+
+The general ExecutionHook API has two parts, spec and status. The hook spec has two piece of information, what are the commands to execute and where to execute them. In many use cases, different hooks share the same commands and user or controller will create many hooks for execution repeatedly. To reduce the work of copying the same execution commands in different hooks, we also propose a second API, HookAction to record the execution commands which can be referenced by ExecutionHook API. Both APIs are namespaced and they should be in the same namespace.
+
+The Create event of ExecutionHook will trigger the hook controller to run HookAction command. The status will be updated once it is available by controller. It is the caller's (user or controller) responsibility to delete the hook once the execution finishes (either succeeded or failed). Otherwise, ExecutionHook controller will garbage collect them after a certain amount of time.
+
+Here is the definition of an ExecutionHook API object:
+
+```
+// ExecutionHook is in the tenant namespace
+type ExecutionHook struct {
+        metav1.TypeMeta
+        // +optional
+        metav1.ObjectMeta
+
+        // Spec defines the behavior of a hook.
+        // +optional
+        Spec ExecutionHookSpec
+
+        // Status defines the current state of a hook.
+        // +optional
+        Status ExecutionHookStatus
+}
+```
+
+Here is the definition of the ExecutionHookSpec:
+
+```
+// HookActionName is copied to ExecutionHookSpec by the controller such as
+// the Snapshot Controller.
+type ExecutionHookSpec struct {
+        // PodSelection defines how to select pods and containers to run
+	// the executionhook. This field is required.
+        PodSelection PodSelection
+
+        // Name of the HookAction. This is required.
+        HookActionName string
```

Comment:
You'll almost certainly need parameters to define policy.

Should it stop if there's a failure?

Should pods be acted upon all in parallel?

Should there be a delay between executions?

Should it retry?  How many times and how fast?

Reply:
before we do have a policy, but remove it for simplicity. The default behavior would be actions would be executed in containers asynchronous. There is no ordering. If the callers requires some ordering,  they have to create executionHook in some order and wait for one to succeed to create the following ones. So executionHook controller's logic would be simplified. The executionHook will retry the execute the action in container within ActionTimeout defined in the spec.

Reply:
OK.  I expect it to come back.